Add sourceMap on tsconfig in order to fix test coverage issues

marianadasilvadev opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

This PR adds sourceMap property on tsconfig file, which will fix some problems related to the frontend test coverage:

- The missed lines were being reported wrong.
- The import statements of Angular files were being missed, dropping all Angular files coverage, even if it was completely tested.

Here's some prints to describe the fix visually:
Before:
![Screenshot from 2019-12-17 07-14-09](https://user-images.githubusercontent.com/34922478/70989597-ab5cd500-20a2-11ea-8e44-26ab5c9fec16.png)
![Screenshot from 2019-12-17 07-14-37](https://user-images.githubusercontent.com/34922478/70989609-af88f280-20a2-11ea-936e-499c6ef317ad.png)

After:
![Screenshot from 2019-12-17 07-53-17](https://user-images.githubusercontent.com/34922478/70989631-b4e63d00-20a2-11ea-886c-69cefc44c3d6.png)
![Screenshot from 2019-12-17 07-52-57](https://user-images.githubusercontent.com/34922478/70989639-b7e12d80-20a2-11ea-862b-2a02132c26c2.png)

## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `python -m scripts.run_frontend_tests`.
- [x] The PR is made from a branch that's **not** called "develop".
- [ ] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [ ] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

Commit:
Add sourceMap on tsconfig in order to fix the karma coverage report,
which the missed lines were wrong, and the import statements from Angular
files, which were being missed.

nithusha21 reviewed (approved):
Thanks for the fix! Nice to know that we got to the bottom of this :)

nithusha21 commented:
@ankita240796 @seanlip @kevinlee12 could you PTAL at this?

seanlip reviewed (approved):
Thanks @marianazangrossi! Well done on finding the error :D

ankita240796 reviewed (approved):
LGTM, Thanks @marianazangrossi for fixing this!

One doubt though - can you clarify a bit that how generating a sourceMap fixes the testing issue? Thanks!

kevinlee12 reviewed (commented):
I remember attempting to fix this coverage issue back in July and I saw explicit instructions not to add the `sourceMap` config and to instead change the karma configuration file because this will tie this to the production code when that get compiled.

I remember attempting to fix this in July, here is PR: https://github.com/oppia/oppia/pull/7138/.

seanlip commented:
@kevinlee12, could you please add a link to the "explicit instructions" you mention?

Also /cc @vojtechjelinek in case he has context.

kevinlee12 commented:
sorry, I don’t have links to those anymore. 

> On Dec 17, 2019, at 10:18 PM, Sean Lip <notifications@github.com> wrote:
> 
> ﻿
> @kevinlee12, could you please add a link to the "explicit instructions" you mention?
> 
> Also /cc @vojtechjelinek in case he has context.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

vojtechjelinek reviewed (changes requested):
I don't have that much context about the source maps, but if it fixes the issue with coverage it is probably okay to merge this.

Although, one thing that should probably be tested is if the server error reporting still works correctly (https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/App.ts#L217; introduced in #7498), because it works with the map files.

/cc @kevintab95

kevinlee12 commented:
Ah, so here it is:

https://stackoverflow.com/a/44787972

`sourceMap` conflicts with this setting https://github.com/oppia/oppia/blob/6d73c3627568dabd0d1b776d914e5e13c9984017/core/tests/karma.conf.ts#L154.

kevinlee12 commented:
Also it’s better to change whatever is in `karma.conf.ts` so that there aren’t any race conditions between that’s set there and `tsconfig.json`.

marianadasilvadev commented:
@kevinlee12 If `karma.conf.ts` will be changed instead, I think it'll need a new loader for webpack, because I've used some karma plugins but it didn't work. Also, I've tested `inlineSourceMap` property in` tsconfig.json` instead of `sourceMap` property and it works. What is the best approach for this case? I didn't quite understand the server error reporting from `App.ts`

kevinlee12 commented:
>  If karma.conf.ts will be changed instead, I think it'll need a new loader for webpack, because I've used some karma plugins but it didn't work

Can you clarify by what you mean by new loader? 

> I didn't quite understand the server error reporting from App.ts

What server error are you getting?

^ After that, I can suggest an approach :)

marianadasilvadev commented:
@kevinlee12 I was thinking of [source-map-loader](https://www.npmjs.com/package/source-map-loader). I don't know if it's gonna work, but webpack loaders is the only solution I can think without changing `tsconfig.json`.

I'm not getting any server error, so I'm bit confused how it works.

kevinlee12 commented:
The other option is to get rid of that line that I had linked and add the corresponding setting in `tsconfig.json`. If that works, lets go with that.

Commit:
Add inlineSourceMap instead of sourceMap in tsconfig.json and change the rules order in wepack

marianadasilvadev commented:
@kevinlee12 For some reason, the `sourceMap` property is very unstable. So I changed it to `inlineSourceMap` and I changed the order of rules in webpack configuration, I think when `ts-loader` was being declared before, it was affecting the testing coverage, but it's just a guess.

kevinlee12 reviewed (approved):
lgtm!

kevinlee12 commented:
code looks good, but the front end tests failed. ptal!

marianadasilvadev commented:
I've runned it and it worked! @kevinlee12

marianadasilvadev commented:
Just a note: I was testing this new config on others branches, in order to actually know which lines were missed and the karma reporter turned weird:
![Screenshot from 2019-12-20 14-49-30](https://user-images.githubusercontent.com/34922478/71279610-f2172d00-2337-11ea-912e-d7fae4721b63.png)

Some files is 0% coverage and I fix it using the `skipFilesWithNoCoverage`, but I'm not sure if this should be fixed (before and after the fix, the total coverage were the same):
```
coverageIstanbulReporter: {
      reports: ['html', 'lcovonly'],
      dir: '../karma_coverage_reports/',
      fixWebpackSourcePaths: true,
      skipFilesWithNoCoverage: true,
      'report-config': {
        html: { outdir: 'html' }
      }
    }, 
```
![Screenshot from 2019-12-20 14-46-35](https://user-images.githubusercontent.com/34922478/71279345-8b920f00-2337-11ea-9efd-c596432bc8cf.png)

I'll test it further before the merge. What do you think about this weird behavior? Should the `skipFilesWithNoCoverage` be added in the `karma.conf.ts`?

seanlip commented:
I don't think you should skip reporting of files with no coverage (assuming I understand what's going on correctly). Those should register as 0%.

seanlip commented:
Also there seems to be a Travis failure that seems unrelated. I'm restarting it, but, for reference, the log is here:

```

**************************************************
*                    Failures                    *
**************************************************
1) Issues visualization records early quit issue.
  - Expected $.length = 5 to equal 0.
  Expected $[0] = Entry({ level: SEVERE, message: 'http://localhost:9001/third_party/static/MathJax-2.7.5/config/TeX-AMS-MML_HTMLorMML.js?V=2.7.5 - Failed to load resource: the server responded with a status of 404 (Not Found)', timestamp: 1576886263954, type: '' }) to equal undefined.
  Expected $[1] = Entry({ level: SEVERE, message: 'http://localhost:9001/third_party/static/MathJax-2.7.5/config/TeX-AMS-MML_HTMLorMML.js?V=2.7.5 - Failed to load resource: the server responded with a status of 404 (Not Found)', timestamp: 1576886267811, type: '' }) to equal undefined.
  Expected $[2] = Entry({ level: SEVERE, message: 'http://localhost:9001/third_party/static/MathJax-2.7.5/config/TeX-AMS-MML_HTMLorMML.js?V=2.7.5 - Failed to load resource: the server responded with a status of 404 (Not Found)', timestamp: 1576886355891, type: '' }) to equal undefined.
  Expected $[3] = Entry({ level: SEVERE, message: 'http://localhost:9001/third_party/static/MathJax-2.7.5/config/TeX-AMS-MML_HTMLorMML.js?V=2.7.5 - Failed to load resource: the server responded with a status of 404 (Not Found)', timestamp: 1576886377235, type: '' }) to equal undefined.
  Expected $[4] = Entry({ level: SEVERE, message: 'http://localhost:9001/third_party/static/MathJax-2.7.5/config/TeX-AMS-MML_HTMLorMML.js?V=2.7.5 - Failed to load resource: the server responded with a status of 404 (Not Found)', timestamp: 1576886380911, type: '' }) to equal undefined.
2) Issues visualization records multiple incorrect issue.
  - Expected $.length = 3 to equal 0.
  Expected $[0] = Entry({ level: SEVERE, message: 'http://localhost:9001/third_party/static/MathJax-2.7.5/config/TeX-AMS-MML_HTMLorMML.js?V=2.7.5 - Failed to load resource: the server responded with a status of 404 (Not Found)', timestamp: 1576886397903, type: '' }) to equal undefined.
  Expected $[1] = Entry({ level: SEVERE, message: 'http://localhost:9001/third_party/static/MathJax-2.7.5/config/TeX-AMS-MML_HTMLorMML.js?V=2.7.5 - Failed to load resource: the server responded with a status of 404 (Not Found)', timestamp: 1576886409968, type: '' }) to equal undefined.
  Expected $[2] = Entry({ level: SEVERE, message: 'http://localhost:9001/third_party/static/MathJax-2.7.5/config/TeX-AMS-MML_HTMLorMML.js?V=2.7.5 - Failed to load resource: the server responded with a status of 404 (Not Found)', timestamp: 1576886412030, type: '' }) to equal undefined.
3) Issues visualization records cyclic transitions issue.
  - Expected $.length = 3 to equal 0.
  Expected $[0] = Entry({ level: SEVERE, message: 'http://localhost:9001/third_party/static/MathJax-2.7.5/config/TeX-AMS-MML_HTMLorMML.js?V=2.7.5 - Failed to load resource: the server responded with a status of 404 (Not Found)', timestamp: 1576886429419, type: '' }) to equal undefined.
  Expected $[1] = Entry({ level: SEVERE, message: 'http://localhost:9001/third_party/static/MathJax-2.7.5/config/TeX-AMS-MML_HTMLorMML.js?V=2.7.5 - Failed to load resource: the server responded with a status of 404 (Not Found)', timestamp: 1576886468262, type: '' }) to equal undefined.
  Expected $[2] = Entry({ level: SEVERE, message: 'http://localhost:9001/third_party/static/MathJax-2.7.5/config/TeX-AMS-MML_HTMLorMML.js?V=2.7.5 - Failed to load resource: the server responded with a status of 404 (Not Found)', timestamp: 1576886471899, type: '' }) to equal undefined.
```

seanlip commented:
@vojtechjelinek @nithusha21 could you PTAL at this PR?

seanlip commented:
(Also @kevinlee12 in case he wants to weigh in on any of the recent comments.)

kevinlee12 commented:
Yeah, as Sean mentioned, let's keep the files that report 0% coverage. 

As far as the the e2e test is concerned, I went ahead and cleared out the cache and restarted the test.

nithusha21 commented:
I agree with Sean and Kevin on the 0% issue. Other than that, I have no concerns.

marianadasilvadev commented:
Thanks! So the e2e tests were just flicky and this PR is not affecting it?

vojtechjelinek reviewed (approved):
LGTM!
